### PR TITLE
Add postprocessing take 3

### LINKED ISSUE #419: Add option to reverse animation
**Is your feature request related to a problem? Please explain.**
I don't know if I can't find it, but it looks like there is no easy way to reverse an animation

**Describe the solution you'd like**
Maybe one could add an argument to the render function so it is changes from 
```
function render(
    video::Video;
    framerate = 30,
    pathname = "javis_$(randstring(7)).gif",
    liveview = false,
    streamconfig::Union{StreamConfig,Nothing} = nothing,
    tempdirectory = "",
    ffmpeg_loglevel = "panic",
    rescale_factor = 1.0,
)
    layers = video.layers
    objects = video.objects
    frames = preprocess_frames!(video)
```
to
```julia
function render(
    video::Video;
    framerate = 30,
    pathname = "javis_$(randstring(7)).gif",
    liveview = false,
    streamconfig::Union{StreamConfig,Nothing} = nothing,
    tempdirectory = "",
    ffmpeg_loglevel = "panic",
    rescale_factor = 1.0,
    reverse_animation = false
)
    layers = video.layers
    objects = video.objects
    frames = preprocess_frames!(video)
    if reverse_animation 
        frames = reverse(frames)
    end
```

**Describe alternatives you've considered**
One could also create a `reverse_render` function, I have no clue which is better.

I can make the PR if it makes sense.

### PR DESCRIPTION
**Pull request by gpucce**
## PR Checklist

If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [x] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [x] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [x] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [ ] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [ ] Did I properly add test dependencies to the `test` directory (if applicable)?
- [ ] Did I check relevant tutorials that may be affected by changes in this PR?
- [x] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #419 

**How did you address these issues with this PR? What methods did you use?**
I add 2 keyword arguments to the `render` function, `postprocess_frames_flow` and `postprocess_frame`. 

The first is called on the `frames` vector and can be used to extend it and manipulate it however one wants. 

The second is called after each `get_javis_frame` call and adds a last postprocessing step. In order to not repeat this call in those cases when the same frame appears more than once in the frames vector I try with this: add a `frames_memory` Dict to act as a memoization for the frames but at each iteration check if the current frame will appear in the remainder of the frames to process and in case it does not remove it from the memory.

### CODE REVIEW

**Comment by gpucce** [comment]
Hi @Wikunia , I tried to do what we said on zulip, that is:
 - move postprocessing in a function
 - move rescaling within the postprocessing function
 - add docstrings to all functions and comments
 - add tests
Let me know what is missing and I will try to add it!

File: src/Javis.jl [hunk 1]
```diff
@@ -233,6 +233,12 @@ Streaming to Twitch or other platforms are not yet supported.
     - Can be used if there are errors with ffmpeg. Defaults to panic:
     All other options are described here: https://ffmpeg.org/ffmpeg.html
 - `rescale_factor::Float64` factor to which the frames should be rescaled for faster rendering
+- `postprocess_frame::Function` function that is applied to the imagematrix of each frame after they have been computed 
+takes as argument postprocess_frame(frame_image, frame, frames), useful to apply a postprocessing e.g. blur to all or some
```
Comment (Wikunia) [inline src/Javis.jl:237]: I would write `frame_image, frame, frames` so without the function name and in the code ticks

File: src/util.jl [hunk 2]
```diff
@@ -260,3 +260,136 @@ Currently only implemented is to change from `<:Integer` to `float`
...
+        else
+            frame_image = _convert_and_rescale_if_needed(
+                get_javis_frame(video, objects, frame; layers = layers),
+                rescale_factor,
+            )
+        end
+    end
+
+    # Memoization container updating with cleaning
+    if !(frame in frames[(filecounter + 1):end]) & haskey(frames_memory, frame)
```
Comment (Wikunia) [inline src/util.jl:382]: We normally use the short circuit operators `&&` and `||`

File: src/util.jl [hunk 3]
```diff
@@ -260,3 +260,136 @@ Currently only implemented is to change from `<:Integer` to `float`
...
+        (sizemargin ÷ 2 + 2):(sizefrom - sizemargin ÷ 2)
+    end
+end
+
+"""
+    crop(im, heightto, widthto)
+
+Crops an imagematrix to size `(heightto, widthto)` around the center of the iamge.
+"""
+function crop(im, heightto, widthto)
```
Comment (Wikunia) [inline src/util.jl:287]: Cropping doesn't seem to be tested.

File: src/util.jl [hunk 4]
```diff
@@ -260,3 +260,136 @@ Currently only implemented is to change from `<:Integer` to `float`
...
+        (sizemargin ÷ 2 + 1):(sizefrom - sizemargin ÷ 2)
+    else
+        (sizemargin ÷ 2 + 2):(sizefrom - sizemargin ÷ 2)
+    end
+end
+
+"""
+    crop(im, heightto, widthto)
+
+Crops an imagematrix to size `(heightto, widthto)` around the center of the iamge.
```
Comment (Wikunia) [inline src/util.jl:285]: *image

File: src/util.jl [hunk 5]
```diff
@@ -260,3 +260,136 @@ Currently only implemented is to change from `<:Integer` to `float`
 """
 interpolateable(x::AbstractVector) = x
 interpolateable(x::AbstractVector{<:Integer}) = float.(x)
+
+
+
+"""
+    _get_range(sizemargin, sizefrom)
+
+For even numbers returns a range with values from `sizemargin ÷ 2 + 1` to `sizefrom - sizemargin ÷ 2`.
```
Comment (Wikunia) [inline src/util.jl:269]: For an even `sizemargin` that is, right?

File: src/Javis.jl [hunk 6]
```diff
@@ -272,20 +280,48 @@ function render(
...
+            frames = frames,
+            rescale_factor = rescale_factor,
+        )
 
         if !isempty(tempdirectory)
             Images.save("$(tempdirectory)/$(lpad(filecounter, 10, "0")).png", frame_image)
         end
         if render_mp4
-            if frame == first(frames)
+            if frame == first(frames) & !started_encoding_video
```
Comment (Wikunia) [inline src/Javis.jl:323]: here also the short circuit `&&`

File: src/Javis.jl [hunk 7]
```diff
@@ -272,20 +280,48 @@ function render(
...
+    # Create a template useful in case of postprocessing
+    frame_template = _convert_and_rescale_if_needed(
+        get_javis_frame(video, objects, first(frames); layers = layers),
+        rescale_factor,
+    )
+
+    # Check if postprocess_frames_flow is ill defined
+    original_n_frames = length(frames)
+    frames = postprocess_frames_flow(frames)
+    if !issubset(unique(frames), 1:original_n_frames)
```
Comment (Wikunia) [inline src/Javis.jl:292]: the `unique` isn't necessary

File: src/Javis.jl (hunk 6 above)
Comment (Wikunia) [reply src/Javis.jl:323]: I'm unsure why we check for both here actually. Can't we just check it with `if !started_encoding_video` ?

File: src/Javis.jl [hunk 8]
```diff
@@ -272,20 +280,48 @@ function render(
...
+        error(
```
Comment (Wikunia) [inline src/Javis.jl:293]: Should be also tested with `@test_throws`

File: src/Javis.jl (hunk 6 above)
Comment (Wikunia) [reply src/Javis.jl:323]: Also just for good measure maybe add a test case where the first frame is not `1` which should make a difference.

**Review by Wikunia — CHANGES_REQUESTED** [review]
Thanks again for the PR @gpucce !
I had some time for a more in depth review. Just small things and some test cases are missing, mainly for cropping.

File: src/Javis.jl (hunk 6 above)
Comment (gpucce) [reply src/Javis.jl:323]: I will add the test! The reason for the check is that if `postprocess_frames_flow` repeats the first frame then the encoding starts twice which I though would be bad, though I don't know VideoIO very much (at all to be fair xD)

**Comment by gpucce** [comment]
I should have done it all. I could only test the returned size for the `crop` function because if I try to make two identical imagematrices one making a small imagematrix and one cropping a large imagematrix they don't match exactly. Though they do look the way they should so maybe this is fine?

File: src/Javis.jl (hunk 6 above)
Comment (Wikunia) [reply src/Javis.jl]: Yeah only checking for the first frame will fail bit only checking whether we initialized rendering should be enough, or not?

Comment (gpucce) [reply src/Javis.jl]: For sure I am not understanding smth, but if after `postprocessing_frames_flow` the vector `frames` is [1,2,1] for example would't it start rendering a second time at the third step without the second check?

Comment (Wikunia) [reply src/Javis.jl]: I would only use the second check 🤣

Comment (gpucce) [reply src/Javis.jl]: ooooooh got I make it now xD

**Comment by gpucce** [comment]
@Wikunia I did the @btime test and looks like nothing changes!!

**Review by Wikunia — APPROVED** [review]
LGTM. Thank you!

**Comment by Sov-trotter** [comment]
This looks cool. 
I didn't take a detailed look at the code but seems like the tests compares the rendered images with itself. There is no reference image. https://github.com/Wikunia/Javis.jl/pull/427/files#diff-9e0faea70cdac333b698a72a30357133df0d1b0bbfcd8251172efecd452a47b2R42-R44

**Comment by Wikunia** [comment]
It tests it with post processing and without postprocessing. So it kind of generates its own reference images all the time. Which is quite reasonable I think for seeing when things should behave exactly like a different version of it.

**Comment by Sov-trotter** [comment]
Ah. I saw that now. Sorry

**Pull request merged**